### DISCUSSION TITLE
Custom timestamp field

### DISCUSSION THREAD

**Discussion by apremalal** [uber-go/zap#1129] [Q&A] [answered]
I'm trying to customise the time stamp field as mentioned in the last comment of this thread.

https://github.com/uber-go/zap/issues/33

```
logger := zap.New(
  zap.NewJSONEncoder(zap.RFC3339Formatter("@timestamp")),
  zap.ErrorLevel,
  zap.Output(someFile),
)
```

But it seems the content is outdated already. I couldn't locate `zap.RFC3339Formatter`  and `zap.NewJSONEncoder` is moved to `zapcore.NewJSONEncoder`. An example usage with the new API changes would be a great thing to have.

**Comment by akshayjshah** [comment]
Those APIs are all extremely outdated - they predate zap's 1.0 release.

Try starting with the [current GoDoc](https://godoc.org/go.uber.org/zap) instead.

**Comment by apremalal** [comment]
Not enough documentation on timestamp usage. I've moved to a different logging library. Thanks.

**Comment by akshayjshah** [comment]
Cool - best of luck!

**Comment by dmuntean** [comment]
For the sake of those who come to this issue trying to configure a custom timestamp, here is an example of how to do it with current API:

	atom := zap.NewAtomicLevel()

	encoderCfg := zap.NewProductionEncoderConfig()
	encoderCfg.TimeKey = "timestamp"
	encoderCfg.EncodeTime = zapcore.ISO8601TimeEncoder

	logger := zap.New(zapcore.NewCore(
		zapcore.NewJSONEncoder(encoderCfg),
		zapcore.Lock(os.Stdout),
		atom,
	))

	defer logger.Sync()

	atom.SetLevel(zap.DebugLevel)

**Comment by grosser** [comment]
FYI smaller way of doing a custom config:
```go
loggerConfig := zap.NewProductionConfig()
loggerConfig.EncoderConfig.TimeKey = "timestamp"
logger, err := loggerConfig.Build()
```

**Comment by gopaltirupur** [comment]
https://blog.sandipb.net/2018/05/03/using-zap-creating-custom-encoders/

**Comment by bytemain** [comment]
use a human-readable time format:
```go
var cfg zap.Config = zap.NewProductionConfig()
// cfg.OutputPaths = []string{"judger.log"}
cfg.EncoderConfig.EncodeTime = zapcore.RFC3339TimeEncoder

logger, err := cfg.Build()
```

**Comment by happybydefault** [comment, marked as answer]
I think this is the best way, as you can use any Go time layout string you want:

```golang
config := zap.NewProductionConfig() // or zap.NewDevelopmentConfig() or any other zap.Config
config.EncoderConfig.EncodeTime = zapcore.TimeEncoderOfLayout(time.RFC3339) // or time.RubyDate or "2006-01-02 15:04:05" or even freaking time.Kitchen

logger, err := config.Build()
```

For more information on how to format the layout string, check https://pkg.go.dev/time#pkg-constants

**Comment by rking-bennie** [comment]
I appreciate you all following up on this thread. TBH I almost left zap as well because I thought this should be simple and virtually every top google link was not helpful. I would assume that everyone who looks to use your framework wants to do this since the default is not that helpful. I would recommend maybe putting this simple example in your README. I've used several framework across several different languages and pretty new to Go but it was painful to finally find this post with the answer i needed

**Comment by sywhang** [comment]
@rking-bennie thank you for the feedback. Appreciate the patience and apologies for the difficulties you ran into. 

Couple of things:
1. I'll turn this thread into a discussion so that it can be more easily discovered.
2. We are discussing internally to gather effort for creating a documentation site that goes over core concepts and HOWTOs for some of the core open sourced frameworks and libraries, and Zap is on top of the mind for us. We will keep you updated via discussions announcement when that effort gets consolidated.

Thanks.

**Comment by plastikfan** [comment]
Please can we have this properly documented. How on earth did such a good library as this, release it with poor representation of timestamp? I am new to zap and the first thing that horified me was the timestamp fiasco and to make things worse, its not clearly documented how to fix it. Good job I stumbled upon this discussion.

**Reply by abhinav** [reply]
Agreed that this could be better documented.
I've raised https://github.com/uber-go/zap/pull/1234 to add more documentation on this topic.
Hopefully, this helps.

**Reply by cschar** [reply]
Just found this discussion through the issue it spawned from , after searching all through https://pkg.go.dev/go.uber.org/zap
looking for documentation how to set an easy ISO8601 timestamp in a production setting.

Yes, please add a doc site!
